source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc

```diff
@@ -0,0 +1,142 @@
+#include "extensions/transport_sockets/proxy_protocol/proxy_protocol.h"
+
+#include <sstream>
+
+#include "envoy/config/core/v3/proxy_protocol.pb.h"
+#include "envoy/network/transport_socket.h"
+
+#include "common/buffer/buffer_impl.h"
+#include "common/network/address_impl.h"
+
+#include "extensions/common/proxy_protocol/proxy_protocol_header.h"
+
+using envoy::config::core::v3::ProxyProtocolConfig_Version;
+using Envoy::Extensions::Common::ProxyProtocol::PROXY_PROTO_V1_AF_INET;
+using Envoy::Extensions::Common::ProxyProtocol::PROXY_PROTO_V1_AF_INET6;
+using Envoy::Extensions::Common::ProxyProtocol::PROXY_PROTO_V1_SIGNATURE;
+using Envoy::Extensions::Common::ProxyProtocol::PROXY_PROTO_V2_ADDR_LEN_INET;
+using Envoy::Extensions::Common::ProxyProtocol::PROXY_PROTO_V2_ADDR_LEN_INET6;
+using Envoy::Extensions::Common::ProxyProtocol::PROXY_PROTO_V2_AF_INET;
+using Envoy::Extensions::Common::ProxyProtocol::PROXY_PROTO_V2_AF_INET6;
+using Envoy::Extensions::Common::ProxyProtocol::PROXY_PROTO_V2_ONBEHALF_OF;
+using Envoy::Extensions::Common::ProxyProtocol::PROXY_PROTO_V2_SIGNATURE;
+using Envoy::Extensions::Common::ProxyProtocol::PROXY_PROTO_V2_SIGNATURE_LEN;
+using Envoy::Extensions::Common::ProxyProtocol::PROXY_PROTO_V2_TRANSPORT_STREAM;
+using Envoy::Extensions::Common::ProxyProtocol::PROXY_PROTO_V2_VERSION;
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace ProxyProtocol {
+
+ProxyProtocolSocket::ProxyProtocolSocket(Network::TransportSocketPtr transport_socket,
+                                         Network::TransportSocketOptionsSharedPtr options,
+                                         ProxyProtocolConfig_Version version)
+    : transport_socket_(std::move(transport_socket)), options_(options), version_(version) {}
+
+void ProxyProtocolSocket::setTransportSocketCallbacks(
+    Network::TransportSocketCallbacks& callbacks) {
+  transport_socket_->setTransportSocketCallbacks(callbacks);
+  callbacks_ = &callbacks;
+}
+
+std::string ProxyProtocolSocket::protocol() const { return transport_socket_->protocol(); }
+
+absl::string_view ProxyProtocolSocket::failureReason() const {
+  return transport_socket_->failureReason();
+}
+
+bool ProxyProtocolSocket::canFlushClose() { return transport_socket_->canFlushClose(); }
+
+void ProxyProtocolSocket::closeSocket(Network::ConnectionEvent event) {
+  transport_socket_->closeSocket(event);
+}
+
+Network::IoResult ProxyProtocolSocket::doRead(Buffer::Instance& buffer) {
+  return transport_socket_->doRead(buffer);
+}
+
+Network::IoResult ProxyProtocolSocket::doWrite(Buffer::Instance& buffer, bool end_stream) {
+  if (header_buffer_.length() > 0) {
+    auto header_res = writeHeader();
+    if (header_buffer_.length() == 0 && header_res.action_ == Network::PostIoAction::KeepOpen) {
+      auto inner_res = transport_socket_->doWrite(buffer, end_stream);
```

Comment:
Do we want to use doWrite here?  I think for if the inner transport socket is TLS, that's going to result in an SSL_write where we want to literally do a raw write to the socket, no?

Reply:
`writeHeader` does the raw write. This is just proxying the `buffer` passed to `doWrite` after the header is written. So I _think_ this is what we want to do here.

Reply:
oh yeah, I totally misread that, my bad

Reply:
Yep, totally my bad misreading this.